In [1]:
import pandas as pd

In [2]:
elections = pd.read_csv("elections.csv")
elections.columns

Index(['Year', 'Candidate', 'Party', 'Popular vote', 'Result', '%'], dtype='object')

In [3]:
import zipfile

zf = zipfile.ZipFile("namesbystate.zip", 'r')
#print(zf.)
ca_name = 'STATE.CA.TXT'
field_names = ['State', 'Sex', 'Year', 'Name', 'Count']
with zf.open(ca_name) as fh:
    babynames = pd.read_csv(fh, header=None, names=field_names)

babynames.head()

State Sex  Year      Name  Count
0    CA   F  1910      Mary    295
1    CA   F  1910     Helen    239
2    CA   F  1910   Dorothy    220
3    CA   F  1910  Margaret    163
4    CA   F  1910   Frances    134

# BASIC SLICING

In [10]:
# VERY BASIC SLICING EXAMPLE

elections.loc[0:4]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

In [11]:
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

In [12]:
elections.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

In [13]:
elections.loc[0:4, "Year":"Party"]

Year          Candidate                  Party
0  1824     Andrew Jackson  Democratic-Republican
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic
3  1828  John Quincy Adams    National Republican
4  1832     Andrew Jackson             Democratic

# Loc

loc selects items by label. First argument is rows, second argument is columns.

Arguments to loc can be:
    
▪ A list.

▪ A slice (syntax is inclusive of the right hand side of the slice).

▪ A single value.

In [14]:
elections.loc[[87, 25, 179], ["Year", "Candidate", "Result"]]

Year             Candidate Result
87   1932        Herbert Hoover   loss
25   1860  John C. Breckinridge   loss
179  2020          Donald Trump   loss

In [15]:
elections.loc[[87, 25, 179], "Popular vote":"%"]

Popular vote Result          %
87       15761254   loss  39.830594
25         848019   loss  18.138998
179      74216154   loss  46.858542

In [16]:
elections.loc[[87, 25, 179], "Popular vote"]

87     15761254
25       848019
179    74216154
Name: Popular vote, dtype: int64

In [17]:
elections.loc[0, "Candidate"]

'Andrew Jackson'

In [18]:
elections.loc[:, ["Year", "Candidate", "Result"]]

Year          Candidate Result
0    1824     Andrew Jackson   loss
1    1824  John Quincy Adams    win
2    1828     Andrew Jackson    win
3    1828  John Quincy Adams   loss
4    1832     Andrew Jackson    win
..    ...                ...    ...
177  2016         Jill Stein   loss
178  2020       Joseph Biden    win
179  2020       Donald Trump   loss
180  2020       Jo Jorgensen   loss
181  2020     Howard Hawkins   loss

[182 rows x 3 columns]

# iloc

Pandas also supports another operator called iloc.
Fundamentally iloc selects items by number. 

▪ Row numbers are 0 through 181 (in this example, same as labels!).

▪ Column numbers are 0 through 5.

Arguments to iloc can be:
    
▪ A list.

▪ A slice (syntax is exclusive of the right hand side of the slice).

▪ A single value.


In [20]:
elections.iloc[[1, 2, 3], [0, 1, 2]]

Year          Candidate                  Party
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic
3  1828  John Quincy Adams    National Republican

In [21]:
elections.iloc[[1, 2, 3], 0:3]

Year          Candidate                  Party
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic
3  1828  John Quincy Adams    National Republican

In [24]:
elections.iloc[[1, 2, 3], 1]

1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
Name: Candidate, dtype: object

In [26]:
elections.iloc[:, 0:3]

Year          Candidate                  Party
0    1824     Andrew Jackson  Democratic-Republican
1    1824  John Quincy Adams  Democratic-Republican
2    1828     Andrew Jackson             Democratic
3    1828  John Quincy Adams    National Republican
4    1832     Andrew Jackson             Democratic
..    ...                ...                    ...
177  2016         Jill Stein                  Green
178  2020       Joseph Biden             Democratic
179  2020       Donald Trump             Republican
180  2020       Jo Jorgensen            Libertarian
181  2020     Howard Hawkins                  Green

[182 rows x 3 columns]

# loc VS iloc

When choosing between loc and iloc, you’ll usually choose loc.

▪ Safer: If the order of columns gets shuffled in a public database, your code still works.
    
▪ Legible: Easier to understand what elections.loc[:, ["Year", "Candidate", "Result"]] means than 
elections.iloc[:, [0, 1, 4]]

iloc can still be useful.

▪ Example: If you have a DataFrame of movie earnings sorted by earnings, can use iloc to get the median 
earnings for a given year (index into the middle).

# []

[] only takes one argument, which may be:
    
▪ A slice of row numbers.

▪ A list of column labels.

▪ A single column label.

In [27]:
elections[3:7]

Year          Candidate                Party  Popular vote Result  \
3  1828  John Quincy Adams  National Republican        500897   loss   
4  1832     Andrew Jackson           Democratic        702735    win   
5  1832         Henry Clay  National Republican        484205   loss   
6  1832       William Wirt         Anti-Masonic        100715   loss   

           %  
3  43.796073  
4  54.574789  
5  37.603628  
6   7.821583

In [28]:
elections[["Year", "Candidate", "Result"]].tail(5)

Year       Candidate Result
177  2016      Jill Stein   loss
178  2020    Joseph Biden    win
179  2020    Donald Trump   loss
180  2020    Jo Jorgensen   loss
181  2020  Howard Hawkins   loss

In [29]:
elections["Candidate"].tail(5)

177        Jill Stein
178      Joseph Biden
179      Donald Trump
180      Jo Jorgensen
181    Howard Hawkins
Name: Candidate, dtype: object

# RETRIEVING ROW AND COLUMN LABELS

In [32]:
elections.index

RangeIndex(start=0, stop=182, step=1)

In [33]:
elections.columns

Index(['Year', 'Candidate', 'Party', 'Popular vote', 'Result', '%'], dtype='object')

# Conditional Selection

In [35]:
babynames_first_10_rows = babynames.loc[:9, :]

In [36]:
babynames_first_10_rows.loc[[True, False, True, False, True, False, True, False, True, False], :]

State Sex  Year      Name  Count
0    CA   F  1910      Mary    295
2    CA   F  1910   Dorothy    220
4    CA   F  1910   Frances    134
6    CA   F  1910    Evelyn    126
8    CA   F  1910  Virginia    101

In [38]:
logical_operator = (babynames["Sex"] == "F")
logical_operator

0          True
1          True
2          True
3          True
4          True
          ...  
407423    False
407424    False
407425    False
407426    False
407427    False
Name: Sex, Length: 407428, dtype: bool

In [39]:
babynames.loc[babynames["Sex"] == "F"]

State Sex  Year      Name  Count
0         CA   F  1910      Mary    295
1         CA   F  1910     Helen    239
2         CA   F  1910   Dorothy    220
3         CA   F  1910  Margaret    163
4         CA   F  1910   Frances    134
...      ...  ..   ...       ...    ...
239532    CA   F  2022    Zemira      5
239533    CA   F  2022     Ziggy      5
239534    CA   F  2022     Zimal      5
239535    CA   F  2022     Zosia      5
239536    CA   F  2022     Zulay      5

[239537 rows x 5 columns]

In [40]:
babynames[(babynames["Sex"] == "F") & (babynames["Year"] < 2000)]

State Sex  Year      Name  Count
0         CA   F  1910      Mary    295
1         CA   F  1910     Helen    239
2         CA   F  1910   Dorothy    220
3         CA   F  1910  Margaret    163
4         CA   F  1910   Frances    134
...      ...  ..   ...       ...    ...
149050    CA   F  1999    Zareen      5
149051    CA   F  1999    Zeinab      5
149052    CA   F  1999     Zhane      5
149053    CA   F  1999      Zoha      5
149054    CA   F  1999     Zoila      5

[149055 rows x 5 columns]

# QUESTION #1

Which of the following pandas statements returns a DataFrame of the first 3 baby names with Count > 250.

In [47]:
baby_names = babynames.loc[babynames["Count"]>250 , ['Name', 'Count']].head(3)
baby_names

Name  Count
0    Mary    295
233  Mary    390
484  Mary    534

In [48]:
# One more Method

baby_names = babynames.iloc[[0, 233, 484], [3, 4]]
baby_names

Name  Count
0    Mary    295
233  Mary    390
484  Mary    534

In [51]:
# One more Method

baby_names = babynames.loc[babynames["Count"] > 250, ["Name", "Count"]].iloc[0:3, :]
baby_names

Name  Count
0    Mary    295
233  Mary    390
484  Mary    534

# EXAM PROBLEMS

In [54]:
babynames[(babynames["Name"] == "Bella") |
(babynames["Name"] == "Alex") |
(babynames["Name"] == "Ani") |
(babynames["Name"] == "Lisa")]

State Sex  Year   Name  Count
6289      CA   F  1923  Bella      5
7512      CA   F  1925  Bella      8
12368     CA   F  1932   Lisa      5
14741     CA   F  1936   Lisa      8
17084     CA   F  1939   Lisa      5
...      ...  ..   ...    ...    ...
393248    CA   M  2018   Alex    495
396111    CA   M  2019   Alex    438
398983    CA   M  2020   Alex    379
401788    CA   M  2021   Alex    333
404663    CA   M  2022   Alex    344

[364 rows x 5 columns]

Pandas provides many alternatives, for example:
    
▪ .isin

▪ .str.startswith

▪ .groupby.filter


In [55]:
names = ["Bella", "Alex", "Ani", "Lisa"]
babynames[babynames["Name"].isin(names)]

State Sex  Year   Name  Count
6289      CA   F  1923  Bella      5
7512      CA   F  1925  Bella      8
12368     CA   F  1932   Lisa      5
14741     CA   F  1936   Lisa      8
17084     CA   F  1939   Lisa      5
...      ...  ..   ...    ...    ...
393248    CA   M  2018   Alex    495
396111    CA   M  2019   Alex    438
398983    CA   M  2020   Alex    379
401788    CA   M  2021   Alex    333
404663    CA   M  2022   Alex    344

[364 rows x 5 columns]

In [56]:
babynames[babynames["Name"].str.startswith("N")]

State Sex  Year    Name  Count
76        CA   F  1910   Norma     23
83        CA   F  1910  Nellie     20
127       CA   F  1910    Nina     11
198       CA   F  1910    Nora      6
310       CA   F  1911  Nellie     23
...      ...  ..   ...     ...    ...
407319    CA   M  2022   Nilan      5
407320    CA   M  2022   Niles      5
407321    CA   M  2022   Nolen      5
407322    CA   M  2022  Noriel      5
407323    CA   M  2022  Norris      5

[12229 rows x 5 columns]

In [59]:
elections.groupby("Year").filter(lambda sf: sf["%"].max() < 45).set_index("Year").sort_index()

Candidate                 Party  Popular vote Result  \
Year                                                                    
1860       Abraham Lincoln            Republican       1855993    win   
1860             John Bell  Constitutional Union        590901   loss   
1860  John C. Breckinridge   Southern Democratic        848019   loss   
1860    Stephen A. Douglas   Northern Democratic       1380202   loss   
1912        Eugene V. Debs             Socialist        901551   loss   
1912      Eugene W. Chafin           Prohibition        208156   loss   
1912    Theodore Roosevelt           Progressive       4122721   loss   
1912          William Taft            Republican       3486242   loss   
1912        Woodrow Wilson            Democratic       6296284    win   
1968        George Wallace  American Independent       9901118   loss   
1968       Hubert Humphrey            Democratic      31271839   loss   
1968         Richard Nixon            Republican      31783783    win   
1992          Andre Marrou           Libertarian        290087   loss   
1992          Bill Clinton            Democratic      44909806    win   
1992              Bo Gritz              Populist        106152   loss   
1992     George H. W. Bush            Republican      39104550   loss   
1992            Ross Perot           Independent      19743821   loss   

              %  
Year             
1860  39.699408  
1860  12.639283  
1860  18.138998  
1860  29.522311  
1912   6.004354  
1912   1.386325  
1912  27.457433  
1912  23.218466  
1912  41.933422  
1968  13.571218  
1968  42.863537  
1968  43.565246  
1992   0.278516  
1992  43.118485  
1992   0.101918  
1992  37.544784  
1992  18.956298

# Handy Utility Functions

In [60]:
bella_count = babynames[babynames["Name"] == "Bella"]["Count"]

In [61]:
babynames.shape

(407428, 5)

In [62]:
babynames.size #rows*columns

2037140

In [64]:
babynames.describe()

Year          Count
count  407428.000000  407428.000000
mean     1985.733609      79.543456
std        27.007660     293.698654
min      1910.000000       5.000000
25%      1969.000000       7.000000
50%      1992.000000      13.000000
75%      2008.000000      38.000000
max      2022.000000    8260.000000

In [65]:
babynames["Sex"].describe()

count     407428
unique         2
top            F
freq      239537
Name: Sex, dtype: object

In [68]:
babynames.sample() # Random Sample

State Sex  Year    Name  Count
571    CA   F  1912  Myrtle     30

In [69]:
babynames.sample(5).iloc[:, 2:] 

Year      Name  Count
400900  2020  Esequiel      7
166185  2004      Anel     13
73448   1975    Hector      5
70457   1974     Eliza     11
23415   1946   Manuela      9

In [80]:
babynames[babynames["Year"] == 2000].sample(4).iloc[:, 2:]

Year     Name  Count
149158  2000   Claire    476
149236  2000    Kiara    260
343725  2000   Cristo     11
151769  2000  Jordann      7

In [83]:
babynames["Name"].value_counts()

Jean         223
Francis      221
Guadalupe    218
Jessie       217
Marion       214
            ... 
Renesme        1
Purity         1
Olanna         1
Nohea          1
Zayvier        1
Name: Name, Length: 20437, dtype: int64

In [84]:
babynames["Name"].unique()

array(['Mary', 'Helen', 'Dorothy', ..., 'Zae', 'Zai', 'Zayvier'],
      dtype=object)

In [85]:
babynames["Name"].sort_values()

366001      Aadan
384005      Aadan
369120      Aadan
398211    Aadarsh
370306      Aaden
           ...   
220691      Zyrah
197529      Zyrah
217429      Zyrah
232167      Zyrah
404544      Zyrus
Name: Name, Length: 407428, dtype: object

# Custom Sorts

In [86]:
babynames[babynames["Year"] == 2021].sort_values("Count", ascending=False)

State Sex  Year     Name  Count
401665    CA   M  2021     Noah   2613
401666    CA   M  2021     Liam   2477
232168    CA   F  2021   Olivia   2402
232169    CA   F  2021     Emma   2179
401667    CA   M  2021    Mateo   2122
...      ...  ..   ...      ...    ...
235674    CA   F  2021    Minha      5
235675    CA   F  2021  Mirabel      5
235676    CA   F  2021    Misty      5
235677    CA   F  2021    Moira      5
404544    CA   M  2021    Zyrus      5

[6547 rows x 5 columns]

In [87]:
babynames.sort_values("Name", ascending=False)

State Sex  Year     Name  Count
404544    CA   M  2021    Zyrus      5
197529    CA   F  2011    Zyrah      5
220691    CA   F  2017    Zyrah      6
232167    CA   F  2020    Zyrah      5
217429    CA   F  2016    Zyrah      5
...      ...  ..   ...      ...    ...
362040    CA   M  2007    Aaden     20
398211    CA   M  2019  Aadarsh      6
366001    CA   M  2008    Aadan      7
369120    CA   M  2009    Aadan      6
384005    CA   M  2014    Aadan      5

[407428 rows x 5 columns]

What if we wanted to find the longest names in California?

▪ Just sorting by name won’t work!


In [88]:
babynames.sort_values("Name", key=lambda x: x.str.len(), ascending=False).head()

State Sex  Year             Name  Count
334166    CA   M  1996  Franciscojavier      8
337301    CA   M  1997  Franciscojavier      5
339472    CA   M  1998  Franciscojavier      6
321792    CA   M  1991  Ryanchristopher      7
327358    CA   M  1993  Johnchristopher      5

# Adding, Modifying, and Removing Columns


Let’s try to solve the sorting problem with different approaches:
    
▪ We will create a temporary column, then sort on it.

▪ Approach 1: Adding a column is easy

In [90]:
# Create a Series of the length of each name

babyname_lengths = babynames["Name"].str.len()
babyname_lengths

0         4
1         5
2         7
3         8
4         7
         ..
407423    7
407424    3
407425    4
407426    6
407427    4
Name: Name, Length: 407428, dtype: int64

In [99]:
# Add a column named "name_lengths" that includes the length of each name

babynames["name_lengths"] = babyname_lengths
babynames.head()

State Sex  Year             Name  Count  name_lengths
334166    CA   M  1996  Franciscojavier      8            15
327472    CA   M  1993  Ryanchristopher      5            15
337301    CA   M  1997  Franciscojavier      5            15
337477    CA   M  1997  Ryanchristopher      5            15
312543    CA   M  1987  Franciscojavier      5            15

In [100]:
babynames = babynames.sort_values(by = "name_lengths", ascending=False)
babynames

State Sex  Year             Name  Count  name_lengths
334166    CA   M  1996  Franciscojavier      8            15
102505    CA   F  1986  Mariadelosangel      5            15
327472    CA   M  1993  Ryanchristopher      5            15
321792    CA   M  1991  Ryanchristopher      7            15
327358    CA   M  1993  Johnchristopher      5            15
...      ...  ..   ...              ...    ...           ...
337950    CA   M  1998               Ty     70             2
238222    CA   F  2022               An      8             2
100885    CA   F  1986               My     16             2
257613    CA   M  1945               Ed     51             2
316798    CA   M  1989               Aj      6             2

[407428 rows x 6 columns]

In [101]:
babynames = babynames.drop("name_lengths", axis = "columns")
babynames

State Sex  Year             Name  Count
334166    CA   M  1996  Franciscojavier      8
102505    CA   F  1986  Mariadelosangel      5
327472    CA   M  1993  Ryanchristopher      5
321792    CA   M  1991  Ryanchristopher      7
327358    CA   M  1993  Johnchristopher      5
...      ...  ..   ...              ...    ...
337950    CA   M  1998               Ty     70
238222    CA   F  2022               An      8
100885    CA   F  1986               My     16
257613    CA   M  1945               Ed     51
316798    CA   M  1989               Aj      6

[407428 rows x 5 columns]

# SORTING BY ARBITRARY FUNCTIONS

In [104]:
def dr_ea_count(string):
    return string.count('dr') + string.count('ea')

# Use `map` to apply `dr_ea_count` to each name in the "Name" column

babynames["dr_ea_count"] = babynames["Name"].map(dr_ea_count)
babynames = babynames.sort_values(by = "dr_ea_count", ascending=False)
babynames

State Sex  Year      Name  Count  dr_ea_count
101976    CA   F  1986  Deandrea      6            3
308131    CA   M  1985  Deandrea      6            3
108731    CA   F  1988  Deandrea      5            3
115957    CA   F  1990  Deandrea      5            3
131029    CA   F  1994  Leandrea      5            3
...      ...  ..   ...       ...    ...          ...
187069    CA   F  2009       Ari     14            0
310403    CA   M  1986       Dru      5            0
252809    CA   M  1937       Mel      6            0
61465     CA   F  1970       Sue     39            0
316798    CA   M  1989        Aj      6            0

[407428 rows x 6 columns]

# GROUPBY.AGG

A groupby operation involves some combination of splitting the object, applying a function, and combining 
the results. 

▪ Calling .groupby() generates DataFrameGroupBy objects → "mini" sub-DataFrames

▪ Each subframe contains all rows that correspond to a particular year

A groupby operation involves some combination of splitting the object, applying a function, and combining 
the results. 

▪ Calling .groupby() generates DataFrameGroupBy objects → "mini" sub-DataFrames

▪ Each subframe contains all rows that correspond to a particular year

▪ Since we can't work directly with DataFrameGroupBy objects, we will use aggregation methods to 
summarize each DataFrameGroupBy object into one aggregated row per subframe.

# QUESTION #2

Find the female baby name whose popularity has fallen the most.

In [128]:
fname = babynames[babynames["Sex"] == "F"]
fname = fname.sort_values(["Year", "Count"], ascending=False)
f = fname.head(1)
f

State Sex  Year    Name  Count  dr_ea_count
235835    CA   F  2022  Olivia   2178            0

OR

In [136]:
female_names = babynames[babynames['Sex'] == 'F']

rtp_df = female_names.groupby('Name').apply(lambda group: group['Count'] / group['Count'].max())
most_fallen_name = rtp_df.groupby('Name').min().idxmin()
most_fallen_rtp = rtp_df.groupby('Name').min().min()

print(f"The female baby name whose popularity has fallen the most is '{most_fallen_name}' with an RTP of {most_fallen_rtp}.")


The female baby name whose popularity has fallen the most is 'Jessica' with an RTP of 0.0007193209610128039.


# Question #3

Who won elections in 2020?

In [15]:
elections.loc[(elections["Year"]==2020) & (elections["Result"]=='win')]

Year     Candidate       Party  Popular vote Result          %
178  2020  Joseph Biden  Democratic      81268924    win  51.311515

# Question #4

Tell me the Party name having highest vote_rate(%) (also for each part)

In [35]:
# Using Lambda Function

ele = elections.sort_values(["%"], ascending = False)
e = ele.groupby('Party').agg(lambda x: x.iloc[0])
e

Year               Candidate  Popular vote Result  \
Party                                                                      
American               1856        Millard Fillmore        873053   loss   
American Independent   1968          George Wallace       9901118   loss   
Anti-Masonic           1832            William Wirt        100715   loss   
Anti-Monopoly          1884         Benjamin Butler        134294   loss   
Citizens               1980          Barry Commoner        233052   loss   
Communist              1932       William Z. Foster        103307   loss   
Constitution           2008           Chuck Baldwin        199750   loss   
Constitutional Union   1860               John Bell        590901   loss   
Democratic             1964          Lyndon Johnson      43127041    win   
Democratic-Republican  1824          Andrew Jackson        151271   loss   
Dixiecrat              1948          Strom Thurmond       1175930   loss   
Farmer–Labor           1920   Parley P. Christensen        265398   loss   
Free Soil              1848        Martin Van Buren        291501   loss   
Green                  2000             Ralph Nader       2882955   loss   
Greenback              1880         James B. Weaver        308649   loss   
Independent            1992              Ross Perot      19743821   loss   
Liberal Republican     1872          Horace Greeley       2834761   loss   
Libertarian            2016            Gary Johnson       4489235   loss   
National Democratic    1896          John M. Palmer        134645   loss   
National Republican    1828       John Quincy Adams        500897   loss   
National Union         1864         Abraham Lincoln       2211317    win   
Natural Law            1996            John Hagelin        113670   loss   
New Alliance           1988           Lenora Fulani        217221   loss   
Northern Democratic    1860      Stephen A. Douglas       1380202   loss   
Populist               1892         James B. Weaver       1041028   loss   
Progressive            1912      Theodore Roosevelt       4122721   loss   
Prohibition            1892            John Bidwell        270879   loss   
Reform                 1996              Ross Perot       8085294   loss   
Republican             1972           Richard Nixon      47168710    win   
Socialist              1912          Eugene V. Debs        901551   loss   
Southern Democratic    1860    John C. Breckinridge        848019   loss   
States' Rights         1956      T. Coleman Andrews        107929   loss   
Taxpayers              1996         Howard Phillips        184656   loss   
Union                  1936           William Lemke        892378   loss   
Union Labor            1888          Alson Streeter        146602   loss   
Whig                   1840  William Henry Harrison       1275583    win   

                               %  
Party                             
American               21.554001  
American Independent   13.571218  
Anti-Masonic            7.821583  
Anti-Monopoly           1.335838  
Citizens                0.270182  
Communist               0.261069  
Constitution            0.152398  
Constitutional Union   12.639283  
Democratic             61.344703  
Democratic-Republican  57.210122  
Dixiecrat               2.412304  
Farmer–Labor            0.995804  
Free Soil              10.138474  
Green                   2.741176  
Greenback               3.352344  
Independent            18.956298  
Liberal Republican     44.071406  
Libertarian             3.307714  
National Democratic     0.969566  
National Republican    43.796073  
National Union         54.951512  
Natural Law             0.118219  
New Alliance            0.237804  
Northern Democratic    29.522311  
Populist                8.645038  
Progressive            27.457433  
Prohibition             2.249468  
Reform                  8.408844  
Republican             60.907806  
Socialist               6.004354  
Southern Democ

In [36]:
# Using Idxmax function

e2 = elections.loc[elections.groupby('Party')['%'].idxmax()]
e2

Year               Candidate                  Party  Popular vote Result  \
22   1856        Millard Fillmore               American        873053   loss   
115  1968          George Wallace   American Independent       9901118   loss   
6    1832            William Wirt           Anti-Masonic        100715   loss   
38   1884         Benjamin Butler          Anti-Monopoly        134294   loss   
127  1980          Barry Commoner               Citizens        233052   loss   
89   1932       William Z. Foster              Communist        103307   loss   
164  2008           Chuck Baldwin           Constitution        199750   loss   
24   1860               John Bell   Constitutional Union        590901   loss   
114  1964          Lyndon Johnson             Democratic      43127041    win   
0    1824          Andrew Jackson  Democratic-Republican        151271   loss   
103  1948          Strom Thurmond              Dixiecrat       1175930   loss   
78   1920   Parley P. Christensen           Farmer–Labor        265398   loss   
15   1848        Martin Van Buren              Free Soil        291501   loss   
155  2000             Ralph Nader                  Green       2882955   loss   
35   1880         James B. Weaver              Greenback        308649   loss   
143  1992              Ross Perot            Independent      19743821   loss   
31   1872          Horace Greeley     Liberal Republican       2834761   loss   
175  2016            Gary Johnson            Libertarian       4489235   loss   
50   1896          John M. Palmer    National Democratic        134645   loss   
3    1828       John Quincy Adams    National Republican        500897   loss   
27   1864         Abraham Lincoln         National Union       2211317    win   
148  1996            John Hagelin            Natural Law        113670   loss   
136  1988           Lenora Fulani           New Alliance        217221   loss   
26   1860      Stephen A. Douglas    Northern Democratic       1380202   loss   
48   1892         James B. Weaver               Populist       1041028   loss   
68   1912      Theodore Roosevelt            Progressive       4122721   loss   
49   1892            John Bidwell            Prohibition        270879   loss   
150  1996              Ross Perot                 Reform       8085294   loss   
120  1972           Richard Nixon             Republican      47168710    win   
66   1912          Eugene V. Debs              Socialist        901551   loss   
25   1860    John C. Breckinridge    Southern Democratic        848019   loss   
110  1956      T. Coleman Andrews         States' Rights        107929   loss   
147  1996         Howard Phillips              Taxpayers        184656   loss   
93   1936           William Lemke                  Union        892378   loss   
42   1888          Alson Streeter            Union Labor        146602   loss   
11   1840  William Henry Harrison                   Whig       1275583    win   

             %  
22   21.554001  
115  13.571218  
6     7.821583  
38    1.335838  
127   0.270182  
89    0.261069  
164   0.152398  
24   12.639283  
114  61.344703  
0    57.210122  
103   2.412304  
78    0.995804  
15   10.138474  
155   2.741176  
35    3.352344  
143  18.956298  
31   44.071406  
175   3.307714  
50    0.969566  
3    43.796073  
27   54.951512  
148   0.118219  
136   0.237804  
26   29.522311  
48    8.645038  
68   27.457433  
49    2.249468  
150   8.408844  
120  60.907806  
66    6.004354  
25   18.138998  
110   0.174883  
147   0.192045  
93    1.960733  
42    1.288861  
11   53.051213

In [5]:
# Using drop_duplicates Function

best_per_party2 = elections.sort_values("%").drop_duplicates(["Party"], keep="last")
best_per_party2

Year               Candidate                  Party  Popular vote Result  \
148  1996            John Hagelin            Natural Law        113670   loss   
164  2008           Chuck Baldwin           Constitution        199750   loss   
110  1956      T. Coleman Andrews         States' Rights        107929   loss   
147  1996         Howard Phillips              Taxpayers        184656   loss   
136  1988           Lenora Fulani           New Alliance        217221   loss   
89   1932       William Z. Foster              Communist        103307   loss   
127  1980          Barry Commoner               Citizens        233052   loss   
50   1896          John M. Palmer    National Democratic        134645   loss   
78   1920   Parley P. Christensen           Farmer–Labor        265398   loss   
42   1888          Alson Streeter            Union Labor        146602   loss   
38   1884         Benjamin Butler          Anti-Monopoly        134294   loss   
93   1936           William Lemke                  Union        892378   loss   
49   1892            John Bidwell            Prohibition        270879   loss   
103  1948          Strom Thurmond              Dixiecrat       1175930   loss   
155  2000             Ralph Nader                  Green       2882955   loss   
175  2016            Gary Johnson            Libertarian       4489235   loss   
35   1880         James B. Weaver              Greenback        308649   loss   
66   1912          Eugene V. Debs              Socialist        901551   loss   
6    1832            William Wirt           Anti-Masonic        100715   loss   
150  1996              Ross Perot                 Reform       8085294   loss   
48   1892         James B. Weaver               Populist       1041028   loss   
15   1848        Martin Van Buren              Free Soil        291501   loss   
24   1860               John Bell   Constitutional Union        590901   loss   
115  1968          George Wallace   American Independent       9901118   loss   
25   1860    John C. Breckinridge    Southern Democratic        848019   loss   
143  1992              Ross Perot            Independent      19743821   loss   
22   1856        Millard Fillmore               American        873053   loss   
68   1912      Theodore Roosevelt            Progressive       4122721   loss   
26   1860      Stephen A. Douglas    Northern Democratic       1380202   loss   
3    1828       John Quincy Adams    National Republican        500897   loss   
31   1872          Horace Greeley     Liberal Republican       2834761   loss   
11   1840  William Henry Harrison                   Whig       1275583    win   
27   1864         Abraham Lincoln         National Union       2211317    win   
0    1824          Andrew Jackson  Democratic-Republican        151271   loss   
120  1972           Richard Nixon             Republican      47168710    win   
114  1964          Lyndon Johnson             Democratic      43127041    win   

             %  
148   0.118219  
164   0.152398  
110   0.174883  
147   0.192045  
136   0.237804  
89    0.261069  
127   0.270182  
50    0.969566  
78    0.995804  
42    1.288861  
38    1.335838  
93    1.960733  
49    2.249468  
103   2.412304  
155   2.741176  
175   3.307714  
35    3.352344  
66    6.004354  
6     7.821583  
150   8.408844  
48    8.645038  
15   10.138474  
24   12.639283  
115  13.571218  
25   18.138998  
143  18.956298  
22   21.554001  
68   27.457433  
26   29.522311  
3    43.796073  
31   44.071406  
11   53.051213  
27   54.951512  
0    57.210122  
120  60.907806  
114  61.344703